finality-aleph/src/sync/handler.rs

```diff
@@ -307,21 +302,111 @@ where
+        let maybe_id = match self
+            .forest
+            .update_justification(justification, maybe_peer)?
+        {
+            true => Some(id),
+            false => None,
+        };
+        self.try_finalize()?;
+        Ok(maybe_id)
+    }
+
+    fn handle_justifications(
+        &mut self,
+        justifications: Vec<J::Unverified>,
+        maybe_peer: Option<I>,
+    ) -> (Option<BlockIdFor<J>>, Option<<Self as HandlerTypes>::Error>) {
+        let mut maybe_id = None;
+        for justification in justifications {
+            maybe_id = match self.handle_justification(justification, maybe_peer.clone()) {
+                Ok(maybe_other_id) => match (&maybe_id, &maybe_other_id) {
+                    (None, _) => maybe_other_id,
+                    (Some(id), Some(other_id)) if other_id.number() > id.number() => maybe_other_id,
+                    _ => maybe_id,
+                },
+                Err(e) => return (maybe_id, Some(e)),
+            };
+        }
+        (maybe_id, None)
+    }
+
+    /// Handle a justification from user returning the action we should take.
+    pub fn handle_justification_from_user(
+        &mut self,
+        justification: J::Unverified,
+    ) -> Result<Option<BlockIdFor<J>>, <Self as HandlerTypes>::Error> {
+        self.handle_justification(justification, None)
+    }
+
+    /// Handle a state response returning the action we should take, and possibly an error.
+    pub fn handle_state_response(
+        &mut self,
+        justification: J::Unverified,
+        maybe_justification: Option<J::Unverified>,
+        peer: I,
+    ) -> (Option<BlockIdFor<J>>, Option<<Self as HandlerTypes>::Error>) {
+        self.handle_justifications(
+            iter::once(justification)
+                .chain(maybe_justification)
+                .collect(),
+            Some(peer),
+        )
+    }
+
+    /// Handle a request response returning the action we should take, and possibly an error.
+    pub fn handle_request_response(
+        &mut self,
+        justifications: Vec<J::Unverified>,
+        headers: Vec<J::Header>,
+        blocks: Vec<B>,
+        peer: I,
+    ) -> (Option<BlockIdFor<J>>, Option<<Self as HandlerTypes>::Error>) {
+        // handle justifications
+        let sync_action = match self.handle_justifications(justifications, Some(peer.clone())) {
+            (sync_action, None) => sync_action,
+            (sync_action, Some(e)) => return (sync_action, Some(e)),
+        };
+
+        // handle headers
+        for header in headers {
+            if let Err(e) = self
+                .forest
+                .update_required_header(&header, Some(peer.clone()))
+            {
+                return (sync_action, Some(Error::Forest(e)));
+            }
+        }
+
+        // handle blocks
+        for block in blocks {
+            match self.forest.importable(&block.header().id()) {
+                true => self.block_importer.import_block(block),
```

Comment:
shouldnt we call `forest.update_body` also here?

Reply:
Not really, we depend on the events to let us know the import really succeeded. In particular this line might fail for multiple reasons.